File: data-structures/LinkedList.js
class Node {
  constructor(val) {
    this.value = val;
    this.next = null;
  }
}

class LinkedList {
  // initialize the linked list with one or multiple values
  constructor(...nodes) {
    this.head = null;
    this.tail = null;
    this.length = 0;

    for (const node of nodes) {
      this.push(node);
      this.length += 1;
    }
  }

  push(value) {
    const newNode = new Node(value);
    // if head is null, set head to this
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length += 1;
  }

  contains(value) {
    let current = this.head;
    while (current) {
      if (current.value === value) {
        return true;
      }
      current = current.next;
    }
    return false;
  }

  addToHead(value) {
    const newNode = new Node(value);
    newNode.next = this.head;
    this.head = newNode;
    this.length += 1;
  }

  insert(value, position) {
    let depth = 0;
    let current = this.head;
    let previous;

    while (depth < position) {
      previous = current;
      current = current.next;
      depth += 1;
    }

    const newNode = new Node(value);
    newNode.next = current;

    if (previous) {
      previous.next = newNode;
    } else {
      this.head = newNode;
    }
    this.length += 1;
  }

  removeItem(value) {
    // remove first occurrence of value from list
    let current = this.head;
    let prev = null;
    while (current) {
      if (current.value === value) {
        if (current === this.head) {
          this.head = this.head.next;
        } else if (current === this.tail) {
          this.tail = prev;
          prev.next = null;
        } else {
          prev.next = current.next;
        }
      }
      prev = current;
      current = current.next;
    }
    this.length -= 1;
  }

  removePosition(position) {
    // remove element at specified position in list
    let depth = 0;
    let current = this.head;
    let previous = current;

    while (depth < position) {
      previous = current;
      current = current.next;
      depth += 1;
    }

    while (current) {
      if (depth === position) {
        // if the node to be removed is the head
        // reassign the head
        if (this.head && current === this.head) {
          this.head = this.head.next;
        }
        // ditto tail
        if (this.tail && current === this.tail) {
          this.tail = previous;
          this.tail.next = null;
        } else {
          // node to be removed is in the middle of list
          previous.next = current.next;
        }
      }
      this.length -= 1;
      return current;
    }
  }

  getLength() {
    return this.length;
  }
}

module.exports = LinkedList;

File: python/data_structures/linked_list.py
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
    
    def append(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self.length += 1
    
    def prepend(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            node.next = self.head
            self.head = node

        self.length += 1
    
    def contains(self, value):
        current = self.head

        while(current):
            if current.value == value:
                return True
            current = current.next

        return False
    
    def remove(self, value):
        current = self.head

        if current.value == value:
            self.head = self.head.next
            self.length -= 1
            return


        while(current.next):
            if current.next.value == value:
                break
            current = current.next

        if not current.next:
            return
        
        if (current.next == self.tail):
            self.tail = current

        current.next = current.next.next
        self.length -= 1

    def remove_at(self, position):
        if position < 0 or position > self.length:
            return

        current = self.head
        # finish
        pass

    def front(self):
        return self.head.value if self.head else None
    
    def back(self):
        return self.tail.value if self.tail else None

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

ll = LinkedList()
ll.append(4)
ll.append(5)
ll.append(6)
ll.append(7)
ll.append(8)

ll.remove(8)
# print(ll.contains(6))
print(ll.contains(8))
print(ll.back())
# print(ll.front())
# print(ll.back())
# print(ll.prepend(1))
# print(ll.front())
# print(ll.contains(5))
# print(ll.contains(6))